src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -138,6 +138,49 @@ public VisitArgumentResult(VisitResult visitResult, Optional<LocalState> stateFo
             }
         }
 
+        private static class PooledDictionaryWithCustomBoundDefaultExpressionComparison
+        {
+            private sealed class CustomEqualityComparer : IEqualityComparer<object>
+            {
+                public static CustomEqualityComparer Instance { get; } = new();
+
+                private CustomEqualityComparer()
+                {
+                }
+
+                public new bool Equals([AllowNull] object x, [AllowNull] object y)
+                {
+                    if (x is BoundDefaultExpression boundDefaultExpression1 &&
+                        y is BoundDefaultExpression boundDefaultExpression2)
+                    {
+                        return boundDefaultExpression1.Type.Equals(boundDefaultExpression2.Type, TypeCompareKind.ConsiderEverything);
```

Comment:
Perhaps it only makes sense to do this tracking if the type in question has `TypeKind == TypeParameter`. In this case we would only do this comparison if we also know the types are type parameters. In this case I think we should also assert that `(object)Type == Type.OriginalDefinition`.

Reply:
Does it matter whether or not we special case type parameters?

Reply:
Have continued to think about this and now believe it *probably* doesn't matter. I would just want to make sure we are not using additional memory to track the state of `default(object)` if the expression is used but never null-tested, for example.